### PR TITLE
Add Basic Auth per Frontend

### LINKED ISSUE #751: HTTP basic auth per frontend
Hi, I think it has been asked before but how would you implement http auth basic on a per backend basis ? Currently it is only on the entrypoint.

### PR DESCRIPTION
**Pull request by SantoDE**
Fixes #751

Currently, only Basic Auth and the docker & rancher provider are supported.

### CODE REVIEW

**Comment by cemo** [comment]
@SantoDE Kubernetes please :)

**Comment by SantoDE** [comment]
Hey @cemo ,

Sadly I've personally used Kubernetes so it's really hard for me to add it. However, if you're experienced in Kubernetes let's meet on Slack ;-)

**Comment by Starefossen** [comment]
Yes, support for Kubernetes would be nice as well 👍

**Comment by joshuacox** [comment]
beautiful!  I will test this out very soon

**Comment by athoune** [comment]
@SantoDE, maybe you can blind code, just with unit tests ;-)

**Comment by seeder** [comment]
Adding to consul_catalog backend would be great

**Comment by Russell-IO** [comment]
Is there any scope to add support for an identify provider ? i.e. LDAP / OAUTH ?

**Comment by ViViDboarder** [comment]
I haven't tested this branch yet, but does it support the functionality added in https://github.com/containous/traefik/pull/889 where the authenticated user is added to the header?

**Comment by OlivierCuyp** [comment]
Seems to be a good first step to me :D

**Comment by kiview** [comment]
Will this work with the file backend as well?

**Comment by scher200** [comment]
@OlivierCuyp awesome men, I look forward to this merge

**Comment by mzuccala** [comment]
@SantoDE Hey, just opened https://github.com/SantoDE/traefik/pull/1 adding support for basic auth in kubernetes :)

**Comment by RXWatcher1** [comment]
@SantoDE Does this allow use of a htpasswd file? Thank you!

**Comment by emilevauge** [comment]
@SantoDE We should move on this one, could you add support for `traefik.<servicename>.*` properties (https://github.com/containous/traefik/pull/1257)?

**Comment by SantoDE** [comment]
Yeah @emilevauge , will do tonight or tommorow :)

**Comment by tha80** [comment]
It's such a shame that this Feature / Pull Request is still not merged, because dynamic basic auth is such an important feature! :(

**Comment by emilevauge** [comment]
@tha80 
>It's such a shame that this Feature / Pull Request is still not merged, because dynamic basic auth is such an important feature! :(

Please, keep in mind that most people are working on this project during their spare time.
Then, Traefik is free, you did not pay anything to use it.
Finally, like all other contributors, you are welcome to create a PR on this :)

**Therefore it's difficult to say that it's "a shame" if a feature is missing...**

**Comment by SantoDE** [comment]
docs updated @vdemeester 👼

**Comment by emilevauge** [comment]
Oh man, @SantoDE is soooo fast !

**Comment by RXWatcher1** [comment]
@SantoDE Are we able to use a htpasswd file with this? Thanks again!

File: docs/toml.md [hunk 1]
```diff
@@ -844,6 +844,7 @@ Labels can be used on containers to override default behaviour:
 - `traefik.frontend.passHostHeader=true`: forward client `Host` header to the backend.
 - `traefik.frontend.priority=10`: override default frontend priority
 - `traefik.frontend.entryPoints=http,https`: assign this frontend to entry points `http` and `https`. Overrides `defaultEntryPoints`.
+- `traefik.frontend.auth.basic=test:$apr1$H6uskkkW$IgXLP6ewTrSuBkTrqE8wj/,test2:$apr1$d9hr9HBB$4HxwgUir3HP4EsggP/QNo0`: Set's a Basic Auth for that frontend with the users test:test and test2:test2
```
Comment (timoreimann) [inline docs/toml.md]: Typo: Set's -> Sets.

File: docs/toml.md [hunk 2]
```diff
@@ -852,11 +853,11 @@ If several ports need to be exposed from a container, the services labels can be
 - `traefik.<service-name>.weight=10`: assign this service weight. Overrides `traefik.weight`.
 - `traefik.<service-name>.frontend.backend=fooBackend`: assign this service frontend to `foobackend`. Default is to assign to the service backend.
 - `traefik.<service-name>.frontend.entryPoints=http`: assign this service entrypoints. Overrides `traefik.frontend.entrypoints`.
+- `traefik.<service-name>.frontend.auth.basic=test:$apr1$H6uskkkW$IgXLP6ewTrSuBkTrqE8wj/,test2:$apr1$d9hr9HBB$4HxwgUir3HP4EsggP/QNo0` Set's a Basic Auth for that frontend with the users test:test and test2:test2.
```
Comment (timoreimann) [inline docs/toml.md]: Sets.

File: docs/toml.md [hunk 3]
```diff
@@ -1601,6 +1602,7 @@ Labels can be used on task containers to override default behaviour:
 - `traefik.frontend.passHostHeader=true`: forward client `Host` header to the backend.
 - `traefik.frontend.priority=10`: override default frontend priority
 - `traefik.frontend.entryPoints=http,https`: assign this frontend to entry points `http` and `https`. Overrides `defaultEntryPoints`.
+- `traefik.frontend.auth.basic=test:$apr1$H6uskkkW$IgXLP6ewTrSuBkTrqE8wj/,test2:$apr1$d9hr9HBB$4HxwgUir3HP4EsggP/QNo0`: Set's a Basic Auth for that frontend with the users test:test and test2:test2
```
Comment (timoreimann) [inline docs/toml.md]: Sets.

**Comment by SantoDE** [comment]
Typos fixed @timoreimann :) So feel free to merge! :D cc: @containous/traefik

**Comment by jri-sp** [comment]
Hi, 
I need this feature too, do you know how many time it usually takes between the merge and the next release ?

Best regards,

**Comment by Starefossen** [comment]
@jri-sp if you look through the [releases](https://github.com/containous/traefik/releases) page it might give you an indication. Hint; look for the time between 2.1 and 2.2.

**Comment by timoreimann** [comment]
@jri-sp @Starefossen previous release cycles are probably not a good indication anymore as we are currently trying to release more frequently. The goal is to ship every other month.

Code freeze is supposed to happen at the end of April. Depending on how much feedback we get during the release candidate phase, 1.3 should be out sooner or later. I'd think on the order of 2-3 weeks.

**Comment by jri-sp** [comment]
@Starefossen @timoreimann 
Alright! thanks both of you for your quick answer!

**Pull request merged**